src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1535,29 +1546,172 @@ julia> vcat(df1, df3, cols=:intersect)
    5 │     8
    6 │     9
 
-julia> vcat(d4, df1)
+julia> vcat(df4, df1)
 3×2 DataFrame
  Row │ A      B
      │ Int64  Int64
 ─────┼──────────────
    1 │     1      1
    2 │     2      2
    3 │     3      3
+
+julia> vcat(df1, df2, df3, df4, cols=:union, source="source")
+9×4 DataFrame
+ Row │ A      B        C        source
+     │ Int64  Int64?   Int64?   Int64
+─────┼─────────────────────────────────
+   1 │     1        1  missing       1
+   2 │     2        2  missing       1
+   3 │     3        3  missing       1
+   4 │     4        4  missing       2
+   5 │     5        5  missing       2
+   6 │     6        6  missing       2
+   7 │     7  missing        7       3
+   8 │     8  missing        8       3
+   9 │     9  missing        9       3
+
+julia> vcat(df1, df2, df4, df3, cols=:union, source=:source => 'a':'d')
+9×4 DataFrame
+ Row │ A      B        C        source
+     │ Int64  Int64?   Int64?   Char
+─────┼─────────────────────────────────
+   1 │     1        1  missing  a
+   2 │     2        2  missing  a
+   3 │     3        3  missing  a
+   4 │     4        4  missing  b
+   5 │     5        5  missing  b
+   6 │     6        6  missing  b
+   7 │     7  missing        7  d
+   8 │     8  missing        8  d
+   9 │     9  missing        9  d
 ```
 """
 Base.vcat(dfs::AbstractDataFrame...;
           cols::Union{Symbol, AbstractVector{Symbol},
-                      AbstractVector{<:AbstractString}}=:setequal) =
-    reduce(vcat, dfs; cols=cols)
+                      AbstractVector{<:AbstractString}}=:setequal,
+          source::Union{Nothing, SymbolOrString,
+                           Pair{<:SymbolOrString, <:AbstractVector}}=nothing) =
+    reduce(vcat, dfs; cols=cols, source=source)
+
+"""
```

Comment:
@nalimilan - I have added a docstring for `reduce` to make this option more discoverable. Could you please have a look at it before I merge? Thank you!